### PR TITLE
bpo-32071: Add unittest -k option

### PR DESCRIPTION
**Pull request by jonashaag**
https://bugs.python.org/issue32071

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Lib/unittest/main.py [hunk 1]
```diff
@@ -225,8 +241,7 @@ def _do_discovery(self, argv, Loader=None):
                 self._initArgParsers()
             self._discovery_parser.parse_args(argv, self)
 
-        loader = self.testLoader if Loader is None else Loader()
-        self.test = loader.discover(self.start, self.pattern, self.top)
+        self.createTests(from_discovery=True, Loader=Loader)
```
Comment (pitrou) [inline Lib/unittest/main.py:244]: I'm curious, why was this change necessary?

File: Lib/unittest/test/test_program.py [hunk 2]
```diff
@@ -409,6 +409,15 @@ def testParseArgsAbsolutePathsThatCannotBeConverted(self):
         # for invalid filenames should we raise a useful error rather than
         # leaving the current error message (import of filename fails) in place?
 
+    def testParseArgsSelectedTestNames(self):
```
Comment (pitrou) [inline Lib/unittest/test/test_program.py:413]: Thank you for adding these tests. Do you think you could also create a small functional test? There is an example of such a test (using `subprocess`) in `Test_TextTestRunner.test_warnings`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for posting this PR. In addition to the review comments I posted, there are two points to address:
* sign the contributor's agreement, if you haven't already done so
* add documentation for the new command-line option and the `TestLoader` addition; the docs are `Doc/library/unittest.rst`

**Comment by warsaw** [comment]
There really should be both an issue opened and a NEWS item for this.  These should not be skipped.  Otherwise, I have to read the diff to infer what `-k` is supposed to do.

**Comment by pitrou** [comment]
I fixed the title to refer to the issue on the bug tracker.
As for the NEWS entry, you're right, I forgot this. @jonashaag, you can add one using the [blurb utility](https://pypi.python.org/pypi/blurb).

**Comment by jonashaag** [comment]
I have made the requested changes; please review again.

File: Lib/unittest/test/test_program.py (hunk 2 above)
Comment (jonashaag) [reply Lib/unittest/test/test_program.py:413]: I added a functional test but I'm not sure if that's the right place to put it.

File: Lib/unittest/main.py (hunk 1 above)
Comment (jonashaag) [reply Lib/unittest/main.py:244]: This refactoring was necessary because I didn't want to repeat the code that sets `loader.testNamePatterns` from the CLI options.  That code has to be executed BEFORE `loader.discover`. Besides, I believe it makes sense to have a single place where `self.test` is set.

Comment (jonashaag) [reply Lib/unittest/main.py]: This refactoring was necessary because I didn't want to repeat the code that sets `loader.testNamePatterns` from the CLI options.  That code has to be executed BEFORE `loader.discover`. Besides, I believe it makes sense to have a single place where `self.test` is set.

**Comment by jonashaag** [comment]
I'm not sure what the Sphinx error means.

File: Doc/library/unittest.rst [hunk 3]
```diff
@@ -1745,6 +1764,21 @@ Loading and running tests
 
       This affects all the :meth:`loadTestsFrom\*` methods.
 
+   .. attribute:: testNamePatterns
+
+      List of Unix shell-style wildcard test name patterns that test methods
+      have to match to be included in test suites (see ``-v`` option).
+
+      If this attribute is not `None` (the default), all test methods to be
```
Comment (Mariatta) [inline Doc/library/unittest.rst]: The sphinx build says:
`[2] library/unittest.rst:1772: default role used`
It points to line 1772 of unittest.rst.
I think it's because single backticks was used around None.
Try using double backticks instead.

File: Doc/library/unittest.rst [hunk 4]
```diff
@@ -229,6 +245,9 @@ Command-line options
 .. versionadded:: 3.5
    The command-line option ``--locals``.
 
+.. versionadded:: 3.7
+  The command-line option ``-k``.
```
Comment (Mariatta) [inline Doc/library/unittest.rst]: Add one more space before "The command-line .. ".

File: Doc/library/unittest.rst [hunk 5]
```diff
@@ -219,6 +219,22 @@ Command-line options
...
+
+   Patterns that contain a wildcard character (``*``) are matched against the
+   test name using :meth:`fnmatch.fnmatchcase`; otherwise simple case-sensitive
+   substring matching is used.
+
+   Patterns are matched against the fully qualified test method name as
+   imported by the test loader.
+
+   For example, ``-k foo`` matches ``foo_tests.SomeTest.test_something``,
+   ``bar_tests.SomeTest.test_foo``, but not `bar_tests.FooTest.test_something``.
```
Comment (Mariatta) [inline Doc/library/unittest.rst]: Mismatched number of backticks in `bar_tests.FooTest.test_something`.

**Comment by jonashaag** [comment]
Not sure why CLA not signed label is still there; I signed it on 2017-11-21.

**Comment by pitrou** [comment]
> Not sure why CLA not signed label is still there; I signed it on 2017-11-21.

Hmm... I've tried to reach to some people who might have a clue about that :-)

**Comment by vstinner** [comment]
> @vstinner vstinner removed the CLA not signed label just now

and

> @the-knights-who-say-ni ay-ni the-knights-who-say-ni added the CLA not signed label just now

@jonashaag: If the bot adds again "not signed", it means that you forgot to fill your GitHub account login in your bugs.python.org profile.

**Comment by vstinner** [comment]
Read also the @the-knights-who-say-ni message which explains that ;-) https://github.com/python/cpython/pull/4496#issuecomment-346157217

**Comment by jonashaag** [comment]
@vstinner It's there, I added it on 2017-11-21.

**Comment by jonashaag** [comment]
Screenshots of signed CLA as received from Adobe sign (?)

Is there a manual check by a human involved in the process? If so, maybe the person simply hasn't found the time to process it.

bildschirmfoto 2017-11-24 um 14 35 59
bildschirmfoto 2017-11-24 um 14 36 09

**Comment by vstinner** [comment]
On bugs.python.org, I still see "Contributor Form Received: No".

Yeah, I think that a human process is required somewhere.

**Comment by ejodlowska** [comment]
@vstinner bugs account has been updated. For whatever reason, the email notification never came through.

**Comment by vstinner** [comment]
Ok, the CLA signed label is now here :-) Thank you Ewa.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
The PR now looks good to me, I'm going to merge.
